### PR TITLE
Add helper methods for registering `StateMachine` states + names

### PR DESCRIPTION
**Pull request by l-Luna**
This adds methods for adding states to `StateMachine`s without using reflection (on `StateMachine` itself or `Seeker`/`AngryOshiro`). It also allows you to give names to states, so mods don't need mod-specific interop for that.

Example usage:
```cs
public static int StSpin;

public override void Load(){
	On.Celeste.Player.ctor += OnPlayerConstruct;
}

public override void Unload(){
	On.Celeste.Player.ctor -= OnPlayerConstruct;
}

private static void OnPlayerConstruct(On.Celeste.Player.orig_ctor orig, Player self, Vector2 position, PlayerSpriteMode spritemode){
	orig(self, position, spritemode);
	StSpin = self.AddState("StateMachineTest:spin", PlayerSpinUpdate);
}

private static int PlayerSpinUpdate(Player p){
	p.Sprite.Rotation += Engine.DeltaTime;
	return StSpin;
}

[Command("smt_start_spin", "Enter StateMachineTest's spin state")]
public static void StartSpinCommand(){
	if(Engine.Scene is not Level scene)
		return;
	Player entity = scene.Tracker.GetEntity<Player>();
	if(entity == null)
		return;

	entity.StateMachine.State = StSpin;
}

[Command("smt_log_state", "Log the player's current state ID & name")]
public static void LogStateCommand(){
	if(Engine.Scene is not Level scene)
		return;
	Player entity = scene.Tracker.GetEntity<Player>();
	if(entity == null)
		return;

	Engine.Commands.Log("state id: " + (int)entity.StateMachine);
	Engine.Commands.Log("state name: " + entity.StateMachine.GetCurrentStateName());
}
```

A few open questions:
- Should vanilla states be given names?
- Originally I wanted to include a way to add "tags" or some metadata to states, so mods could ask e.g. "should I treat this state like a feather, dash, booster, dream dash..." uniformly. But there's already ways of checking specific information (like dash attacking) that's more relevant, and it added a lot of complexity. Would anything similar still be wanted?
- Currently, you still need to hook constructors to add states; for `Seeker`/`AngryOshiro`, you need to make sure you hook the right one. Should an event be added to let you avoid hooks entirely?
- Is the generic version of `AddState` necessary? Are any of the specialized versions necessary?

### CODE REVIEW

**Comment by vivianlonging** [comment]
My personal opinion:
Vanilla States should be given names according to their constant names
The major issue with adding DashState type detection is it could break simple compatibility with vanilla objects. If you wanted to do dash states, _maybe_ adding something like a Player.IsInDashingState would work? But then it becomes a hook management problem, which sucks. So I would say leaving it is a safe bet for at least the first pass, but we could add it in a future push.
I dont think you should add AddState to Player/Seeker/AngryOshiro at all and just only have StateMachine.AddState. However, I think a StateAdd event would be a nice thing to have, except for the concerns of people breaking stuff as a result of easy access.

**Comment by vivianlonging** [comment]
Hi can we please move this forward I kinda need this lol

**Comment by Kalobi** [comment]
Forcing build :wideGladeline:

**Comment by balt-dev** [comment]
image

**Comment by microlith57** [comment]
Screenshot_20230905-174518

**Comment by vivianlonging** [comment]
Please ignore these two screenshots, the PR was not the issue

**Comment by l-Luna** [comment]
thanks, very cool

i'll add the vanilla state names when i'm free, hopefully soon

**Comment by microlith57** [comment]
i've experimented a bit with vanilla state names,  unfortunately the way the machines are added is inconsistent between the three relevant classes, so separate patch code is needed

**Comment by l-Luna** [comment]
alright, here's an updated version! all vanilla states now have names, and all vanilla entities that can be given states now have events for registering them. here's a short mod that demonstrates adding a state and querying existing ones:

```cs
using Monocle;
using MonoMod.Utils;

namespace Celeste.Mod.StateMachineTest;

public class StateMachineTestModule : EverestModule{

	public static int StSpin;

	public override void Load() => Everest.Events.Player.OnRegisterStates += RegisterStates;

	public override void Unload() => Everest.Events.Player.OnRegisterStates -= RegisterStates;

	private static void RegisterStates(Player player) => StSpin = player.AddState("StateMachineTest:spin", PlayerSpinUpdate);

	private static int PlayerSpinUpdate(Player p){
		p.Sprite.Rotation += Engine.DeltaTime;
		return StSpin;
	}

	[Command("smt_start_spin", "Enter StateMachineTest's spin state")]
	public static void StartSpinCommand(){
		if(Engine.Scene is not Level scene)
			return;
		Player entity = scene.Tracker.GetEntity<Player>();
		if(entity == null)
			return;

		entity.StateMachine.State = StSpin;
	}

	[Command("smt_log_state", "Log the player's current state ID & name")]
	public static void LogStateCommand(){
		if(Engine.Scene is not Level scene)
			return;
		Player entity = scene.Tracker.GetEntity<Player>();
		if(entity == null)
			return;

		Engine.Commands.Log("state id: " + (int)entity.StateMachine);
		Engine.Commands.Log("state name: " + entity.StateMachine.GetCurrentStateName());
	}
	
	[Command("smt_log_oshiro_state", "Log Oshiro's current state ID & name")]
	public static void LogOshiroStateCommand(){
		if(Engine.Scene is not Level scene)
			return;
		AngryOshiro entity = scene.Tracker.GetEntity<AngryOshiro>();
		if(entity == null)
			return;

		StateMachine state = DynamicData.For(entity).Get<StateMachine>("state");
		Engine.Commands.Log("state id: " + state.State);
		Engine.Commands.Log("state name: " + state.GetCurrentStateName());
	}
	
	[Command("smt_log_seeker_state", "Log a seeker's current state ID & name")]
	public static void LogSeekerStateCommand(){
		if(Engine.Scene is not Level scene)
			return;
		Seeker entity = scene.Tracker.GetEntity<Seeker>();
		if(entity == null)
			return;

		StateMachine state = DynamicData.For(entity).Get<StateMachine>("State");
		Engine.Commands.Log("state id: " + state.State);
		Engine.Commands.Log("state name: " + state.GetCurrentStateName());
	}
}
```

and as before, `AddState` supports everything that `StateMachine.SetCallbacks` allows, not *just* update functions.

File: Celeste.Mod.mm/Patches/Monocle/Engine.cs [hunk 1]
```diff
@@ -59,12 +58,16 @@ public patch_Engine(int width, int height, int windowWidth, int windowHeight, st
...
+            if (scene == null)
+                return 1f;
+
+            float result = 1f;
+            foreach (TimeRateModifier trm in scene.Tracker.GetComponents<TimeRateModifier>()) {
+                if (trm.Enabled)
+                    result *= trm.Multiplier;
+            }
+
+            return result;
```
Comment (JaThePlayer) [inline Celeste.Mod.mm/Patches/Monocle/Engine.cs:70]: uhhh, this seems unrelated to the PR? It is a good change and I approve of it regardless.

File: Celeste.Mod.mm/Mod/Everest/Everest.cs [hunk 2]
```diff
@@ -315,6 +315,10 @@ public static partial class Everest {
                     if (Enum.TryParse(queue.Dequeue(), ignoreCase: true, out LogLevel level))
                         Logger.SetLogLevelFromSettings("", level);
                 }
+                
+                else if (arg == "--use-scancodes") {
+                    Environment.SetEnvironmentVariable("FNA_KEYBOARD_USE_SCANCODES", "1");
+                }
```
Comment (JaThePlayer) [inline Celeste.Mod.mm/Mod/Everest/Everest.cs:321]: Also unrelated to the PR, is this from some other PR?

File: Celeste.Mod.mm/Mod/Module/EverestModuleMetadata.cs [hunk 3]
```diff
@@ -88,7 +88,7 @@ public sealed class EverestModuleMetadata {
         /// Perform a few basic post-parsing operations. For example, make the DLL path absolute if the mod is in a directory.
         /// </summary>
         public void PostParse() {
-            if (!string.IsNullOrEmpty(DLL) && !string.IsNullOrEmpty(PathDirectory) && !File.Exists(DLL))
+            if (!string.IsNullOrEmpty(DLL) && !string.IsNullOrEmpty(PathDirectory))
```
Comment (JaThePlayer) [inline Celeste.Mod.mm/Mod/Module/EverestModuleMetadata.cs:91]: Also unrelated to the PR

File: Celeste.Mod.mm/Patches/Audio.cs [hunk 4]
```diff
@@ -33,8 +33,11 @@ static class patch_Audio {
         private static HashSet<string> ingestedModBankPaths = new HashSet<string>();
         public static bool AudioInitialized { get; private set; } = false;
 
-        [MonoModIgnore]
-        internal static extern void CheckFmod(RESULT result);
+        [MonoModReplace]
+        internal static void CheckFmod(RESULT result) {
+            if (result != RESULT.OK)
+                throw new Exception($"FMOD Failed: {result} ({Error.String(result)})");
+        }
```
Comment (JaThePlayer) [inline Celeste.Mod.mm/Patches/Audio.cs:40]: Another one

**Review by JaThePlayer — COMMENTED** [review]
Found some stuff that's probably from other PR's, but everything else looks good

**Comment by l-Luna** [comment]
ah, i tried rebasing on top of dev and i assume that's where those came from, i'll fix that

**Review by microlith57 — APPROVED** [review]

**Comment by microlith57** [comment]
Moving to #680 as that PR has a clean git history.

**Comment by microlith57** [comment]
Moving to #680 as that PR has a clean git history.
